mempool/src/shared_mempool/network.rs

```diff
@@ -103,38 +103,66 @@ impl<NetworkClient: NetworkClientInterface<MempoolSyncMsg>> MempoolNetworkInterf
+    fn determine_peers_to_add_and_disable(
+        &self,
+        updated_peers: &HashMap<PeerNetworkId, PeerMetadata>,
+    ) -> (Vec<(PeerNetworkId, ConnectionMetadata)>, Vec<PeerNetworkId>) {
+        let sync_states = self.sync_states.read();
+        let to_disable: Vec<_> = sync_states
+            .keys()
+            .filter(|previous_peer| !updated_peers.contains_key(previous_peer))
+            .copied()
+            .collect();
+        let to_add: Vec<_> = updated_peers
+            .iter()
+            .filter(|(peer, _)| !sync_states.contains_key(peer))
+            .map(|(peer, metadata)| (*peer, metadata.get_connection_metadata()))
+            .filter(|(peer, metadata)| self.is_upstream_peer(peer, Some(metadata)))
+            .collect();
+        (to_add, to_disable)
+    }
+
+    /// Returns newly added peers
+    fn add_and_disable_peers(
+        &self,
+        to_add: &[(PeerNetworkId, ConnectionMetadata)],
+        to_disable: &[PeerNetworkId],
+    ) {
         let mut sync_states = self.sync_states.write();
-        let is_new_peer = !sync_states.contains_key(&peer);
-        if self.is_upstream_peer(&peer, Some(&metadata)) {
-            // If we have a new peer, let's insert new data, otherwise, let's just update the current state
-            if is_new_peer {
+        for (peer, metadata) in to_add.iter().cloned() {
+            if self.is_upstream_peer(&peer, Some(&metadata)) {
                 counters::active_upstream_peers(&peer.network_id()).inc();
                 sync_states.insert(
                     peer,
                     PeerSyncState::new(metadata, self.mempool_config.broadcast_buckets.len()),
                 );
-            } else if let Some(peer_state) = sync_states.get_mut(&peer) {
-                peer_state.metadata = metadata;
             }
         }
-        drop(sync_states);
-
-        // Always need to update the prioritized peers, because of `is_alive` state changes
-        self.update_prioritized_peers();
-        is_new_peer
+        for peer in to_disable {
+            // All other nodes have their state immediately restarted anyways, so let's free them
+            if sync_states.remove(peer).is_some() {
+                counters::active_upstream_peers(&peer.network_id()).dec();
+            }
+        }
     }
 
-    /// Disables a peer if it can be restarted, otherwise removes it
-    pub fn disable_peer(&self, peer: PeerNetworkId) {
-        // All other nodes have their state immediately restarted anyways, so let's free them
-        if self.sync_states.write().remove(&peer).is_some() {
-            counters::active_upstream_peers(&peer.network_id()).dec();
+    /// Update peers based on updated view of connected peers. Return (peers newly added that need
+    /// to start broadcasts, peers that will be disabled from broadcasts).
+    pub fn update_peers(
+        &self,
+        updated_peers: &HashMap<PeerNetworkId, PeerMetadata>,
```

Comment:
nit: does it make sense for this to be called something like "all_connected_peers"?